src/app/pages/send-tokens/components/send-form-inner.tsx

```diff
@@ -67,6 +71,44 @@ export function SendFormInner(props: SendFormInnerProps) {
     values.recipient,
   ]);
 
+  useEffect(() => {
+    void fetchMandatoryMemoAddresses();
+  }, []);
+
+  useEffect(() => {
+    if (mandatoryMemoAddresses.includes(values.recipient)) {
+      setIsMemoRequired(true);
+      return;
+    }
+    setIsMemoRequired(false);
+  }, [values.recipient, mandatoryMemoAddresses]);
+
+  async function fetchMandatoryMemoAddresses() {
+    try {
+      const response = await fetch('https://stacksonchain.com/api/v2/run', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
+          query: `select sender_address as address
+            from transactions tx
+            join stxop.accounts bal on (account = sender_address)
+            --where tx_type = 'token transfer'
+            where type_id = 0
+            group by 1
+            having count(*) >= 100 and count(distinct token_transfer_recipient_address) >= 10
+            limit 100
+          `,
+        }),
+      });
+      const { columns } = await response.json();
+      setMandatoryMemoAddresses(columns.address);
+    } catch (e) {
+      logger.error('sendForm error', e);
+    }
+  }
```

Comment:
Interesting approach using a query to assume exchanges. I wonder if this might false positive some active users, though? Could the query also contain a check for txs with memos?

@markmhx @314159265359879 thoughts on doing this manually vs with a query? 

Rather than declaring this fetch in the component, we typically follow the pattern of delcaring query hooks in `src/app/query`

Reply:
I think if you add a check for memo's to the query you could limit false positives. For example check if more than 50% of the distinct token transfers have a memo. Then even if it is not an exchange we can be quiet sure a memo is still required (example: voting for Stacks 2.1, send a 0.000001 STX transfer with a "yes" or "no" in the memo). 

The message could be: "You are sending to an address that likely requires a memo."

How much time would it take to do these queries, if it takes a lot of time to do the query (>seconds) then hard coding the addresses may be preferred for better user experience?

Address to consider for hard-coding (from stacksonchain probable inflow CEX address'):
<html><body>
<!--StartFragment-->

address | Balance (STX) | xfers | users | memos | Volume (STX)
-- | -- | -- | -- | -- | --
SP1P72Z3704VMT3DMHPP2CB8TGQWGDBHD3RPR9GZS | Ӿ 478,469.888 | 72493 | 31087 | 28613 | Ӿ 659,912,420.965
SP3HXJJMJQ06GNAZ8XWDN1QM48JEDC6PP6W3YZPZJ | Ӿ 19,063,625.052 | 37291 | 6630 | 6291 | Ӿ 384,251,995.968
SPNBYP1MY456K29804XHT4PY5QKMSXNRBHGADTDY | Ӿ 961,716.645 | 11620 | 3005 | 2606 | Ӿ 151,089,780.868
SP3RFAZMSH6YA1KTJD7DN997AG2DG54E3Z9ZJWYN8 | Ӿ 2,359,418.144 | 5122 | 1373 | 1492 | Ӿ 151,046,166.241
SPX8T06E8FJQ33CX8YVR9CC6D9DSTF6JE0Y8R7DS | Ӿ 308.5 | 24467 | 13695 | 10112 | Ӿ 60,801,805.172
SP3SPE6W7VTGM622ZDBYKJ26XQ2R5VN9DDQXFYSVR | Ӿ 213,143.056 | 130 | 99 | 0 | Ӿ 50,190,071.732
SPC3T2E3MXHA15Q1TRQJDH57QX6X1D5D7Y9XTXXJ | Ӿ 2,344,755.586 | 2195 | 2010 | 1 | Ӿ 33,918,875.764
SP1FNEJ21BMJCJYEJM2F73QT6WD2MN9MB2J92FZBN | Ӿ 1,563,130.088 | 2336 | 1909 | 1 | Ӿ 29,490,107.468
SP33XEHK2SXXH625VG6W6665WBBPX1ENQVKNEYCYY | Ӿ 1,338,530.129 | 5070 | 3966 | 8 | Ӿ 25,420,514.732
SP1PJWKS5V3X7JR300J3W291DDNCKD6M6XKJFYPAC | Ӿ 1,596,795.212 | 2243 | 1910 | 1 | Ӿ 23,515,874.648
SPXJJ6XPRV52JREBENWFX7TG1CFPFP0QS81BDH0A | Ӿ 9,451,848.895 | 2934 | 143 | 487 | Ӿ 18,840,698.537
SPFPBQ30DZ2G4CX8XZGZXD013N2NPHH7K7SB7GEZ | Ӿ 1,566,830.034 | 1186 | 991 | 1 | Ӿ 14,407,755.391
SM21NR8W96KGG4YCFXRHZ6EAR2PAY3NS0MQ3FK95T | Ӿ 1,711.333 | 132 | 132 | 0 | Ӿ 12,291,711.333
SP1RWW86QN0KKZYZAZ5K4NJX4BQQCVWVSR7CXW5XA | Ӿ 3,521,122.377 | 2611 | 1071 | 1312 | Ӿ 10,367,625.875
SP335ZC3W6M308ZKTSCCBMR54GWA20RSXCKJPZC59 | Ӿ 21,179.24 | 395 | 22 | 1 | Ӿ 5,553,602.684
SP1Q9S8J351ZMSV3SAHB88CRP9MD2X71WHSAR83AB | Ӿ 255,892.525 | 2065 | 1783 | 4 | Ӿ 5,349,627.457
SP111MNWTSXGTD0ESMV59WX4KHQA93RTV9F82EK0K | Ӿ 4,950,240.466 | 347 | 142 | 208 | Ӿ 4,888,817.241
SP1VZB1N7YFBRNX4CC0WAYWY5MHEXYXYF56QH4F1A | Ӿ 7.444 | 368 | 11 | 1 | Ӿ 4,107,016.799
SM468VETKA5DB15HWG2QM7Y04EFQKV44R9D6D0QC | Ӿ 109,151.348 | 159 | 90 | 87 | Ӿ 3,463,169.046
SP32QTYYGG6SWTP198FST4SPM85J0A3JPNB9S2BEA | Ӿ 428.768 | 124 | 19 | 2 | Ӿ 2,704,762.835
SP27ANV45PCAG98PGFA2GVN9K7QYY1KWWS1V6RFSX | Ӿ 72.9 | 802 | 483 | 458 | Ӿ 1,694,393.993
SP2VK9TBG8E20A0YW228PC70GBMSBFHE7BNVMKB57 | Ӿ 30,619.054 | 2553 | 997 | 905 | Ӿ 1,366,110.57
SP2GHPX871GG50BE72CNFZZ0FJM5JQ5GPDKM4T2RC | Ӿ 6.554 | 274 | 11 | 9 | Ӿ 1,220,835.097
SPEJYD1YTYKK0RBQH200TV7X7ZJC8KMAA5D3SXX3 | Ӿ 450,003.159 | 152 | 35 | 0 | Ӿ 1,142,493.577
SPYABDY1GR8RPQVNK1VFD1CQP6MZT2GF4Q9QB19F | Ӿ 4.186 | 609 | 493 | 168 | Ӿ 1,085,583.678
SP2BE8YZF117DZX8FTZQ0072GQ8VET2PPPQC58YFJ | Ӿ 7.44 | 354 | 305 | 5 | Ӿ 986,441.208
SP3C9JPPGYZV04QF3J0N3ME7KN5NDVF79AZ96HR1 | Ӿ 24,621.471 | 203 | 17 | 3 | Ӿ 900,747.308
SP22GF51PH2HRMWME1HJMDB54VFY5YMZXCMXKZ26V | Ӿ 2,577.003 | 432 | 339 | 349 | Ӿ 724,054.913
SP3Z7511VWR5WG9J3MAKER3NRZYKWT83K2XTP36EV | Ӿ 147.579 | 209 | 71 | 24 | Ӿ 672,544.6
SP3ATSKXC6M2DMN001W20D2GSNEJDX1NXP6N85JEP | Ӿ 11,337.557 | 321 | 194 | 14 | Ӿ 615,130.459
SM3DGPERE6T07F1TN1BAR01P62SWYJDAAAE7YYCQ | Ӿ 1,454.852 | 217 | 31 | 116 | Ӿ 530,381.396
SP2ST741GVK200RYK9NKTYP6CJH1AG4ZPV4A5WG4M | Ӿ 1.419 | 326 | 262 | 2 | Ӿ 482,351
SP1G01K04TVDFY635HSTHNWNQSEWSR5CZF12KAXG4 | Ӿ 1.028 | 109 | 21 | 1 | Ӿ 468,940.927
SP2J02RVJPTKVN81YF9TFPGWEJPHMA79VS5TWGGVK | Ӿ 0 | 329 | 323 | 0 | Ӿ 445,605.445
SP24K6H0FE6Q2QDKAB2YZPE220TE534SPM288D59B | Ӿ 213,587.521 | 123 | 10 | 3 | Ӿ 391,945.23
SP12E5REG9ZPB91V61A8V4P8XJ63EK3VT058YWZAS | Ӿ 6,072.073 | 122 | 13 | 4 | Ӿ 357,276.665
SP3J9CPJJQW12Y08MDSQZDX81YMB40QCSZ1AMVFQV | Ӿ 111,118.756 | 2497 | 2303 | 1 | Ӿ 308,441.936
SPH7P46DMTTFNSHDD33N6G8GFE1MQB2R0QW176B1 | Ӿ 101,564.84 | 308 | 45 | 6 | Ӿ 293,036.177
SPZ1WQQ3R94QT4T5FP2WE5HA7B8MGGH4VFAVB9NY | Ӿ 78,470.603 | 345 | 306 | 0 | Ӿ 265,833.25
SP2HVP68NY5BD2RDFX0JNXSYRS8AA6R7S30N08NJZ | Ӿ 92.189 | 507 | 25 | 1 | Ӿ 235,082.968
SP1H45JS07GWQWMT57JE20X17AQCNVYAS7NHW2HVR | Ӿ 157.066 | 128 | 31 | 5 | Ӿ 220,456.041
SP18XC4F27VQ8P2QGKZ5P6KR41GK77ZVFWV468P1 | Ӿ 10.003 | 174 | 139 | 1 | Ӿ 199,859
SP8TP159Z6V3MN5YGXYV6D5KRDH7D28TYVNJ2QHR | Ӿ 341.409 | 168 | 29 | 17 | Ӿ 198,025.969
SP3M6D6M2BS7FNEFV111ZF6WQYATNJZ89Q7MXSPAE | Ӿ 48,454.479 | 109 | 25 | 2 | Ӿ 151,291.1
SP16CNZGZ20Y85GMWG01A9GYJ3ZHN4XFQJMSTEG8 | Ӿ 2.75 | 167 | 14 | 25 | Ӿ 145,217.529
SPV48Q8E5WP4TCQ63E9TV6KF9R4HP01Z8WS3FBTG | Ӿ 193.612 | 133 | 17 | 3 | Ӿ 85,380.892
SPDXC0NM3YQDHV1HN3V9P5Y4P26QWY709NB86EYB | Ӿ 301.736 | 158 | 21 | 3 | Ӿ 75,315.377
SP329G766AV8Z01X9EEAHPDQ4WDJXT2A0XB383MGP | Ӿ 2.033 | 500 | 155 | 5 | Ӿ 61,654.856
SP78Q12M26WVN1V9DPQ29HVDTWPKQH6KVR1X0VEW | Ӿ 5.574 | 152 | 124 | 0 | Ӿ 61,406
SP3Z0BHS0SVP5733GDZ2RWJV1G2EW04PFN2NSX3PK | Ӿ 16.172 | 265 | 48 | 21 | Ӿ 58,238.101
SP3T7WAB5DMJ3JSRMCQF6SC7CG50DYYJVS4C303CN | Ӿ 10,101.15 | 110 | 19 | 3 | Ӿ 58,022.76
SPV00QHST52GD7D0SEWV3R5N04RD4Q1PMA3TE2MP | Ӿ 28.807 | 161 | 66 | 8 | Ӿ 57,010.737
SP2BBZNE2TZ5A56P3094JKTQCWCT13R61F8B8GXH8 | Ӿ 1.428 | 366 | 77 | 1 | Ӿ 54,357.053
SP2F2KYSKAMT47ZGC4J1R2WB8E0GBF9GG088NJ66X | Ӿ 1.677 | 164 | 19 | 6 | Ӿ 52,144.829
SP2DY54XRX9GQEYXM5KM9RNE28BR7X4GKGC9DXSEA | Ӿ 43,561.863 | 339 | 337 | 0 | Ӿ 50,767.97
SP3E545ADCKY56EVCXZPA87525VM0ZA8DQQAEP77Z | Ӿ 625.754 | 232 | 157 | 1 | Ӿ 47,433.304
SP3Y6P2B33Y448WGCV6VR8CA8GCDXY6QBY5ZBZ13B | Ӿ 26.787 | 403 | 399 | 1 | Ӿ 45,162.246
SP2GD8MA2BVR0DJ3S6GW9BSXY3Q74YR9YS3JXKZDQ | Ӿ 0 | 439 | 439 | 1 | Ӿ 43,869.091
SP1X34E47XW77TWYFG6GX8G1SEN5E538T0WGBXZK6 | Ӿ 4.061 | 115 | 75 | 0 | Ӿ 43,858.592
SP248HH800501WYSG7Z2SS1ZWHQW1GGH85ME34NT2 | Ӿ 141.6 | 352 | 127 | 323 | Ӿ 38,540.16
SP2FHPJ3GMWM31MEQJSDWJTHA00QWMNPYMRZ56SPG | Ӿ 7,360.997 | 3968 | 3957 | 0 | Ӿ 38,092.704
SP081R49TDSQDXFSR8ZE6TAX6T96Z3B80TXNDV7T | Ӿ 8.5 | 113 | 106 | 0 | Ӿ 36,609.926
SPM1Q7YG18378H6W254YN8PABEVRPT38ZCY01SJD | Ӿ 94.547 | 123 | 19 | 8 | Ӿ 34,131.364
SP213KNHB5QD308TEESY1ZMX1BP8EZDPG4JWD0MEA | Ӿ 0.25 | 115 | 64 | 9 | Ӿ 33,327.664
SP9MZHJH0FQB746YZ7D22ZBHJFQVBYN8M8FQ4PCX | Ӿ 0 | 135 | 22 | 2 | Ӿ 28,778.394
SP32CF0E78JNPK0HYDTH3CCZ8FN76PFX5W0FYBN20 | Ӿ 82.532 | 154 | 62 | 6 | Ӿ 28,396.826
SPBN3J1JST1K4MEZXZMV95ZXNBGMMKFZGJKY7VVC | Ӿ 0.994 | 180 | 26 | 1 | Ӿ 27,010.885
SP5NZFW1Z845XY8TX8VP60P0KEC3FRBQCDPBTFCP | Ӿ 0 | 129 | 129 | 0 | Ӿ 25,679.691
SP1WK5VPABSEEA9QSRGTV22YJDZRVHEW0NYERDZX9 | Ӿ 0 | 243 | 243 | 0 | Ӿ 24,299.956
SP59YR7W1J8MRD1D21TE2YJCSXQ9FBT7SFARXFVB | Ӿ 0.148 | 109 | 10 | 1 | Ӿ 23,960.798
SP1XS9E70E81FGEEDMSXCEEPQ2XY94TNVFV1CRJVP | Ӿ 0.099 | 179 | 33 | 0 | Ӿ 21,963.618
SP32DFA3HXYZ2BV3P8H6XQM8EN94D2212QM71BRYG | Ӿ 635.384 | 146 | 127 | 3 | Ӿ 20,564.913
SP1NQ5AS213JFGAS78MF6BPRZ4Z6TENR59K96M67W | Ӿ 27,975.642 | 120 | 118 | 2 | Ӿ 18,708.192
SP3EYT7KF5ERWQFTWW3SWHS8QRYBNSMRZ7JW73YXR | Ӿ 2.176 | 114 | 20 | 1 | Ӿ 17,897.2
SP6RNRKT2QVEN9NXE66MDYXWB01SFNHMW9CVEM8T | Ӿ 0 | 176 | 176 | 0 | Ӿ 17,598.968
SP3297DBXPDGV2FZM8Z77FWA4RE55GY2QBS1CEN08 | Ӿ 0 | 137 | 20 | 2 | Ӿ 17,565.829
SPQE3J7XMMK0DN0BWJZHGE6B05VDYQRXRMDV734D | Ӿ 0.889 | 103 | 20 | 1 | Ӿ 17,037.502
SPQP9F9YV90Y1MDXYF0VDBATAKZ21ER29KTEYPTC | Ӿ 0 | 146 | 146 | 0 | Ӿ 14,699.974
SP267TMY9MTMP98AW05X8RFQ43JRMTPGDWFKFCPTH | Ӿ 0.135 | 178 | 38 | 0 | Ӿ 13,835.497
SP136MDGWQTPYQ9CV89100ZB1MPMVNJEW3DDJMAE2 | Ӿ 0 | 134 | 134 | 0 | Ӿ 13,300.342
SP2NDK60R7JKQ3SJ98CEHV2CMNDVTFBR541C4KV5Y | Ӿ -0.03 | 104 | 44 | 3 | Ӿ 12,766.438
SP1FPE0M7WYP0NCBX9FT5BJB2H5NRBTZ7Q27Y1RT0 | Ӿ 2,816.614 | 104 | 101 | 1 | Ӿ 12,748.265
SP3WPZN2MN4MCESAFGBNX9YZD2JV1RJQRNMYZMMK6 | Ӿ 0.15 | 151 | 149 | 0 | Ӿ 12,720.503
SP1SGWYFVJGHJDVHP90MM89W563DBQGK5N1QJ52T | Ӿ 36.254 | 132 | 15 | 1 | Ӿ 12,403.46
SPWR1705P8D1MWPE8VMESS4AHNSQ26DPJEV5KA1F | Ӿ 0 | 118 | 118 | 0 | Ӿ 11,799.979
SP35MEYYBHSFCFXY296YGP7NAT6Y4XBJW2VETR8AV | Ӿ 3.36 | 136 | 118 | 1 | Ӿ 11,768.967
SP2MDEE7BMXWTNST6PKE8MGP2EWD6412ZNPTYMQ5S | Ӿ 63.963 | 108 | 15 | 3 | Ӿ 11,418.329
SP16BTCDFS05EXR64GBM7SSMSSM4ZF4DVJ3B5JWHZ | Ӿ 0 | 112 | 112 | 0 | Ӿ 11,199.98
SPHM225DX5RXDPK630EDVRWDFBFG750MFDXVHKD4 | Ӿ 0 | 109 | 109 | 0 | Ӿ 10,899.98
SP1J6WAPR86TA1YYJ2JV3HRT7YPKK7APHSZFK6FFK | Ӿ 0 | 108 | 108 | 0 | Ӿ 10,799.981
SP11M5XSC1C37PJ3BC6NEW81ND2PR5SA5Q0C8NS11 | Ӿ 0.169 | 148 | 32 | 0 | Ӿ 7,682.216
SP3EQ7FQ8TFXB792P7VAGRXER0YNFMGM1Y8RS69RC | Ӿ 1,054.256 | 171 | 86 | 4 | Ӿ 6,036.005
SP1J58J25ZY6J9Z07EVHWXWJCNFGBRPPWYNM7W2WD | Ӿ 0 | 122 | 27 | 5 | Ӿ 5,353.974
SP1SS16QBKNXYNPH49RB8HJTJK8J2Q3X3AE55BKFB | Ӿ 0.818 | 225 | 33 | 39 | Ӿ 4,351.46
SP2W0KJMJB2601KK53Y7F8W9FV5YJ1QVCT0GBJHTA | Ӿ 117.74 | 139 | 52 | 1 | Ӿ 3,812.115
SP16JC5P00ZWKMS1YYBTWA9VX8ZQD1750HPFF3MTX | Ӿ 0.535 | 115 | 108 | 0 | Ӿ 3,361.93
SP36SKCSK6DA9D7NSAJATWX4XDVDP7994HH5CCTNB | Ӿ 2,831.261 | 266 | 266 | 0 | Ӿ 3,290.484
SP28C7M3C8A3JV77XKKM8CCVT1S0D8H8CRBTP5BYB | Ӿ 0.117 | 103 | 103 | 0 | Ӿ 2,714.4
SP2SSDNMJYWREZ4T0SRK9DEDDD5DMT094CBB87GSJ | Ӿ 1.834 | 383 | 379 | 33 | Ӿ 2,410.666
SP1JFTR7XH4BCYDD8XVXJ54P4HA1SJ87X1DWCM8SH | Ӿ 0.055 | 196 | 196 | 0 | Ӿ 1,744.639

<!--EndFragment-->
</body>
</html>

Reply:
This approach to querying Stacks on Chain is clever, and I agree could be improved if we were able to query for just those with a predominance of existing (inbound) transactions with memos.

Since false positives are still possible, I'd suggest we take the advisory approach (e.g. displaying a message `⚠️ Important: You may need to include a memo when sending to this address. If you don't, your deposit may fail to get processed by the recipient successfully` while keeping the memo field optional) rather than an obligatory one where we (perhaps erroneously) make it required.

We _could_ hardcode as well, though the querying approach seems more longer-term robust to me (especially if we query in a non-blocking way while loading the send form, in that the warning message will show up afterwards and if for some reason the query takes forever, the user isn't blocked from proceeding to send).

cc @mica000 in case she has input on the design approach here as well

Reply:
several fixes are made:
- query is in query hook
- now there is a default hardcoded array of addresses, in which queried addresses are pushed as soon as they are fetched, if any error happens it doesn't affect send tokens flow
- memo is not mandatory, but if recipient address matches any from the address list, warning and placeholder change (you can see it in the video)

https://user-images.githubusercontent.com/46521087/217333701-23aa8209-d317-432a-a461-a04a705f9b2d.mov